`test_steal_twice` flaky

On Windows, 3.8: https://github.com/dask/distributed/pull/4925/checks?check_run_id=2889177384#step:10:2191

<details><summary>The traceback is just an asyncio.TimeoutError</summary>

```python
______________________________ test_steal_twice _______________________________

    def test_func():
        result = None
        workers = []
        with clean(timeout=active_rpc_timeout, **clean_kwargs) as loop:
    
            async def coro():
                with dask.config.set(config):
                    s = False
                    for _ in range(60):
                        try:
                            s, ws = await start_cluster(
                                nthreads,
                                scheduler,
                                loop,
                                security=security,
                                Worker=Worker,
                                scheduler_kwargs=scheduler_kwargs,
                                worker_kwargs=worker_kwargs,
                            )
                        except Exception as e:
                            logger.error(
                                "Failed to start gen_cluster: "
                                f"{e.__class__.__name__}: {e}; retrying",
                                exc_info=True,
                            )
                            await asyncio.sleep(1)
                        else:
                            workers[:] = ws
                            args = [s] + workers
                            break
                    if s is False:
                        raise Exception("Could not start cluster")
                    if client:
                        c = await Client(
                            s.address,
                            loop=loop,
                            security=security,
                            asynchronous=True,
                            **client_kwargs,
                        )
                        args = [c] + args
                    try:
                        future = func(*args)
                        if timeout:
                            future = asyncio.wait_for(future, timeout)
                        result = await future
                        if s.validate:
                            s.validate_state()
                    finally:
                        if client and c.status not in ("closing", "closed"):
                            await c._close(fast=s.status == Status.closed)
                        await end_cluster(s, workers)
                        await asyncio.wait_for(cleanup_global_workers(), 1)
    
                    try:
                        c = await default_client()
                    except ValueError:
                        pass
                    else:
                        await c._close(fast=True)
    
                    def get_unclosed():
                        return [c for c in Comm._instances if not c.closed()] + [
                            c
                            for c in _global_clients.values()
                            if c.status != "closed"
                        ]
    
                    try:
                        start = time()
                        while time() < start + 60:
                            gc.collect()
                            if not get_unclosed():
                                break
                            await asyncio.sleep(0.05)
                        else:
                            if allow_unclosed:
                                print(f"Unclosed Comms: {get_unclosed()}")
                            else:
                                raise RuntimeError("Unclosed Comms", get_unclosed())
                    finally:
                        Comm._instances.clear()
                        _global_clients.clear()
    
                    return result
    
>           result = loop.run_sync(
                coro, timeout=timeout * 2 if timeout else timeout
            )

distributed\utils_test.py:966: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\Miniconda3\envs\dask-distributed\lib\site-packages\tornado\ioloop.py:530: in run_sync
    return future_cell[0].result()
distributed\utils_test.py:925: in coro
    result = await future
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

fut = <Task cancelled name='Task-157632' coro=<test_steal_twice() done, defined at D:\a\distributed\distributed\distributed\tests\test_steal.py:657>>
timeout = 30

    async def wait_for(fut, timeout, *, loop=None):
        """Wait for the single Future or coroutine to complete, with timeout.
    
        Coroutine will be wrapped in Task.
    
        Returns result of the Future or coroutine.  When a timeout occurs,
        it cancels the task and raises TimeoutError.  To avoid the task
        cancellation, wrap it in shield().
    
        If the wait is cancelled, the task is also cancelled.
    
        This function is a coroutine.
        """
        if loop is None:
            loop = events.get_running_loop()
        else:
            warnings.warn("The loop argument is deprecated since Python 3.8, "
                          "and scheduled for removal in Python 3.10.",
                          DeprecationWarning, stacklevel=2)
    
        if timeout is None:
            return await fut
    
        if timeout <= 0:
            fut = ensure_future(fut, loop=loop)
    
            if fut.done():
                return fut.result()
    
            await _cancel_and_wait(fut, loop=loop)
            try:
                fut.result()
            except exceptions.CancelledError as exc:
                raise exceptions.TimeoutError() from exc
            else:
                raise exceptions.TimeoutError()
    
        waiter = loop.create_future()
        timeout_handle = loop.call_later(timeout, _release_waiter, waiter)
        cb = functools.partial(_release_waiter, waiter)
    
        fut = ensure_future(fut, loop=loop)
        fut.add_done_callback(cb)
    
        try:
            # wait until the future completes or the timeout
            try:
                await waiter
            except exceptions.CancelledError:
                if fut.done():
                    return fut.result()
                else:
                    fut.remove_done_callback(cb)
                    # We must ensure that the task is not running
                    # after wait_for() returns.
                    # See https://bugs.python.org/issue32751
                    await _cancel_and_wait(fut, loop=loop)
                    raise
    
            if fut.done():
                return fut.result()
            else:
                fut.remove_done_callback(cb)
                # We must ensure that the task is not running
                # after wait_for() returns.
                # See https://bugs.python.org/issue32751
                await _cancel_and_wait(fut, loop=loop)
>               raise exceptions.TimeoutError()
E               asyncio.exceptions.TimeoutError

C:\Miniconda3\envs\dask-distributed\lib\asyncio\tasks.py:501: TimeoutError
```

</details>

I ran this test many times per environment between 6 and 17 aug over 3 mass stress test runs.

It failed 38 times out of 66 runs on Windows 3.7.
It failed only once out of 512 runs on all other environments combined.

The error  I'm getting is, invariably,
```
        assert a.in_flight_tasks == 0
>       assert b.in_flight_tasks == 0
E       assert 1 == 0
E         +1
E         -0
```

--- a/distributed/tests/test_steal.py
+++ b/distributed/tests/test_steal.py
@@ -4,7 +4,6 @@ import asyncio
 import contextlib
 import itertools
 import logging
-import math
 import random
 import weakref
 from collections import defaultdict
@@ -112,7 +111,6 @@ async def test_steal_cheap_data_slow_computation(c, s, a, b):
     assert abs(len(a.data) - len(b.data)) <= 5
 
 
-@pytest.mark.slow
 @gen_cluster(
     client=True,
     nthreads=[("", 1)] * 2,
@@ -227,6 +225,7 @@ async def test_stop_in_flight(c, s, a, b):
     assert len(b.state.tasks) != 0
 
 
+@pytest.mark.slow
 @gen_cluster(
     client=True,
     nthreads=[("127.0.0.1", 1)] * 2,
@@ -388,6 +387,7 @@ async def test_dont_steal_fast_tasks_blocklist(c, s, a):
                 assert ts.who_has == {ws_a}
 
 
+@pytest.mark.slow
 @gen_cluster(
     client=True,
     nthreads=[("", 1)],
@@ -414,6 +414,7 @@ async def test_new_worker_steals(c, s, a):
         assert b.data
 
 
+@pytest.mark.slow
 @gen_cluster(
     client=True,
     config={"distributed.scheduler.work-stealing-interval": "100ms"},
@@ -845,48 +846,54 @@ async def test_restart(c, s, a, b):
     assert not any(x for L in steal.stealable.values() for x in L)
 
 
+@pytest.mark.slow
 @gen_cluster(
     client=True,
-    config={"distributed.scheduler.work-stealing-interval": "100ms"},
+    nthreads=[("", 1)],
+    config={
+        # Ensure tasks with dependencies are never rootish
+        "distributed.scheduler.rootish-taskgroup": 0,
+        "distributed.scheduler.work-stealing-interval": "50ms",
+    },
 )
-async def test_steal_twice(c, s, a, b):
-    x = c.submit(inc, 1, workers=a.address)
+async def test_steal_twice(c, s, a):
+    x = c.submit(inc, 1)
     await wait(x)
 
-    futures = [c.submit(slowadd, x, i, delay=0.2) for i in range(100)]
+    futures = [c.submit(slowadd, x, i, delay=0.2) for i in range(120)]
 
-    while len(s.tasks) < 100:  # tasks are all allocated
-        await asyncio.sleep(0.01)
-    if math.isinf(s.WORKER_SATURATION):
+    await async_poll_for(lambda: len(s.tasks) == 121)
+    assert not s.queued  # Thanks to rootish-taskgroup: 0 in the config above
+    await async_poll_for(lambda: len(a.state.tasks) == 121)
+
+    # Add a second worker
+    async with Worker(s.address, nthreads=1) as b:
         # Wait for b to start stealing tasks
-        while len(b.state.tasks) < 30:
-            await asyncio.sleep(0.01)
-    else:
-        # Wait for b to complete some tasks
-        while len(b.data) < 8:
-            await asyncio.sleep(0.01)
+        await async_poll_for(lambda: len(b.state.tasks) >= 30)
+        # Army of new workers arrives to help
+        async with contextlib.AsyncExitStack() as stack:
+            workers = [
+                stack.enter_async_context(Worker(s.address, nthreads=1))
+                for _ in range(10)
+            ]
+            workers = await asyncio.gather(*workers)
 
-    # Army of new workers arrives to help
-    async with contextlib.AsyncExitStack() as stack:
-        # This is pretty timing sensitive
-        workers = [stack.enter_async_context(Worker(s.address)) for _ in range(10)]
-        workers = await asyncio.gather(*workers)
+            await wait(futures)
 
-        await wait(futures)
+            # Note: this includes a and b
+            ntasks = [len(ws.has_what) for ws in s.workers.values()]
+            # Verify that not all tasks stayed on the original workers, and that all
+            # workers had their fare share. With 120 tasks, 12 workers, and 1 thread per
+            # worker, a perfect distribution would imply 10 tasks per worker. A
+            # _reasonable_ distribution means no worker should have a disproportionately
+            # large share and no worker should be left with nothing.
+            assert min(ntasks) > 5, ntasks
+            # Note `async_poll_for(lambda: len(b.state.tasks) >= 30)`
+            # above for the no-queue use case.
+            assert max(ntasks) < 15, ntasks
 
-        # Note: this includes a and b
-        empty_workers = [ws for ws in s.workers.values() if not ws.has_what]
-        assert (
-            len(empty_workers) < 3
-        ), f"Too many workers without keys ({len(empty_workers)} out of {len(s.workers)})"
-        # This also tests that some tasks were stolen from b
-        # (see `while len(b.state.tasks) < 30` above)
-        # If queuing is enabled, then there was nothing to steal from b,
-        # so this just tests the queue was balanced not-terribly.
-        assert max(len(ws.has_what) for ws in s.workers.values()) < 30
-
-        assert a.state.in_flight_tasks_count == 0
-        assert b.state.in_flight_tasks_count == 0
+            assert a.state.in_flight_tasks_count == 0
+            assert b.state.in_flight_tasks_count == 0
 
 
 @gen_cluster(
@@ -1052,6 +1059,7 @@ async def test_parse_stealing_interval(s, interval, expected):
         assert s.periodic_callbacks["stealing"].callback_time == expected
 
 
+@pytest.mark.slow
 @gen_cluster(
     client=True,
     config={"distributed.scheduler.work-stealing-interval": "100ms"},
@@ -1417,7 +1425,6 @@ def test_steal_worker_state(ws_with_running_task):
     assert ws.available_resources == {"R": 1}
 
 
-@pytest.mark.slow()
 @gen_cluster(nthreads=[("", 1)] * 4, client=True)
 async def test_steal_very_fast_tasks(c, s, *workers):
     np = pytest.importorskip("numpy")

--- a/distributed/tests/test_steal.py
+++ b/distributed/tests/test_steal.py
@@ -32,7 +32,6 @@ from distributed.client import Future
 from distributed.compatibility import LINUX
 from distributed.core import Status
 from distributed.metrics import time
-from distributed.system import MEMORY_LIMIT
 from distributed.utils import wait_for
 from distributed.utils_test import (
     NO_AMM,
@@ -639,23 +638,20 @@ async def test_dont_steal_few_saturated_tasks_many_workers(c, s, a, *rest):
 
 @gen_cluster(
     client=True,
-    nthreads=[("127.0.0.1", 1)] * 10,
-    worker_kwargs={"memory_limit": MEMORY_LIMIT},
+    nthreads=[("", 1)],
+    worker_kwargs={"memory_limit": "16 GiB"},
     config={
-        "distributed.scheduler.default-task-durations": {"slowidentity": 0.2},
+        # Ensure tasks with dependencies are never rootish
+        "distributed.scheduler.rootish-taskgroup": 0,
         "distributed.scheduler.work-stealing-interval": "20ms",
     },
 )
-async def test_steal_when_more_tasks(c, s, a, *rest):
-    x = c.submit(mul, b"0", 50000000, workers=a.address)  # 50 MB
-    await wait(x)
-
-    futures = [c.submit(slowidentity, x, pure=False, delay=0.2) for i in range(20)]
-
-    start = time()
-    while not any(w.state.tasks for w in rest):
-        await asyncio.sleep(0.01)
-        assert time() < start + 1
+async def test_steal_large_dependency(c, s, a):
+    x = c.submit(mul, b"0", 50000000)  # 50 MB
+    futures = [c.submit(slowidentity, x, pure=False, delay=0.2) for _ in range(20)]
+    await async_poll_for(lambda: len(a.state.tasks) == 21)
+    async with Worker(s.address, nthreads=1) as b:
+        await async_poll_for(lambda: b.state.tasks)
 
 
 @gen_cluster(

--- a/distributed/tests/test_steal.py
+++ b/distributed/tests/test_steal.py
@@ -884,13 +884,14 @@ async def test_steal_twice(c, s, a):
             ntasks = [len(ws.has_what) for ws in s.workers.values()]
             # Verify that not all tasks stayed on the original workers, and that all
             # workers had their fare share. With 120 tasks, 12 workers, and 1 thread per
-            # worker, a perfect distribution would imply 10 tasks per worker. A
-            # _reasonable_ distribution means no worker should have a disproportionately
-            # large share and no worker should be left with nothing.
-            assert min(ntasks) > 5, ntasks
-            # Note `async_poll_for(lambda: len(b.state.tasks) >= 30)`
-            # above for the no-queue use case.
-            assert max(ntasks) < 15, ntasks
+            # worker, a reasonable distribution means no worker should have a
+            # disproportionately large share and no worker should be left with nothing.
+            # ntasks is routinely observed to be in the [10, 15] range. Give it some
+            # generous margin. Note that you'll never get a perfect distribution of 10
+            # tasks per worker as a and b were started before the other workers.
+            assert min(ntasks) > 3, sorted(ntasks)
+            # Note `async_poll_for(lambda: len(b.state.tasks) >= 30)` above
+            assert max(ntasks) < 22, sorted(ntasks, reverse=True)
 
             assert a.state.in_flight_tasks_count == 0
             assert b.state.in_flight_tasks_count == 0

